src/v/utils/tracking_allocator.cc

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "utils/tracking_allocator.h"
+
+#include "utils/human.h"
+
+#include <sstream>
+
+namespace util {
+
+ss::shared_ptr<mem_tracker> mem_tracker::create_child(ss::sstring label) {
+    auto child = ss::make_shared<mem_tracker>(label);
+    _children.push_back(*child);
+    return child;
+}
+
+// Consumption of self and all children.
+int64_t mem_tracker::consumption() const {
+    int64_t total = _consumption;
+    for (const auto& tracker : _children) {
+        total += tracker.consumption();
+    }
+    return total;
+}
+
+ss::sstring mem_tracker::debug(int level) const {
+    std::stringstream result;
+    for (int i = 0; i <= level; i++) {
+        result << "| ";
+    }
+    result << "\n";
+    for (int i = 0; i < level; i++) {
+        result << "| ";
+    }
+    result << _label << ": " << human::bytes(static_cast<double>(consumption()))
+           << '\n';
+    for (const auto& tracker : _children) {
+        result << tracker.debug(level + 1);
```

Comment:
Nit: This could result in a pretty deep stack depending on the depth of the tracker tree. One way to avoid this is  by representing the call stack explicitly with something like `std::dequeue<std::pair<std::stringstream, decltype(_children::const_interator)>>` to avoid the overhead of a callstack and function calls.

 This would complicate the logic significantly over the recursive version though. So if we don't think the tracker tree will ever get very deep then its probably not worth it.

Reply:
Done, switched to iterative versions in all methods. Not too complex actually.